cmd/serve/s3/utils.go

```diff
@@ -0,0 +1,113 @@
+package s3
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/Mikubill/gofakes3"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/vfs"
+)
+
+func getDirEntries(prefix string, fs *vfs.VFS) (vfs.Nodes, error) {
+	node, err := fs.Stat(prefix)
+
+	if err == vfs.ENOENT {
+		return nil, gofakes3.ErrNoSuchKey
+	} else if err != nil {
+		return nil, err
+	}
+
+	if !node.IsDir() {
+		return nil, gofakes3.ErrNoSuchKey
+	}
+
+	dir := node.(*vfs.Dir)
+	dirEntries, err := dir.ReadDirAll()
+	if err != nil {
+		return nil, err
+	}
+
+	return dirEntries, nil
+}
+
+func getFileHashByte(node interface{}) []byte {
+	b, err := hex.DecodeString(getFileHash(node))
+	if err != nil {
+		return nil
+	}
+	return b
+}
+
+func getFileHash(node interface{}) string {
+	var o fs.Object
+
+	switch b := node.(type) {
+	case vfs.Node:
+		o = b.DirEntry().(fs.Object)
```

Comment:
Note that `o` may be `nil` here.

This happens when a file is in the process of being uploaded but hasn't made it to the storage yet.

In fact I worked around this exact problem in `rclone serve sftp` here 7fc573db27f89e788692f90b6a9dae9b1d35d936